File: laravel/src/X402Middleware.php
<?php

declare(strict_types=1);

namespace X402\Laravel;

use Closure;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;
use X402\Laravel\Contracts\PlatformClientInterface;

/**
 * Route middleware: `->middleware('x402:0.10')`. No business/settlement
 * logic — relays to the platform's signed challenge/verify and fails
 * closed. $next runs ONLY when the platform explicitly allows the request.
 *
 * $price is a telemetry hint (the middleware parameter); the authoritative
 * price is the merchant route registered on the platform.
 */
final class X402Middleware
{
    public function __construct(
        private ?Config $cfg = null,
        private ?PlatformClientInterface $client = null,
    ) {
    }

    public function handle(Request $request, Closure $next, ?string $price = null): mixed
    {
        $cfg = $this->cfg ?? Config::resolve();
        $client = $this->client ?? new PlatformClient($cfg);

        $route = '/' . ltrim($request->path(), '/');
        $method = $request->method();
        $fail = static fn (): JsonResponse => new JsonResponse(['error' => 'x402_platform_unavailable'], 502);

        $proofNonce = $request->header('X-Payment-Nonce');
        if (! $proofNonce) {
            $ch = $client->challenge(['route' => $route, 'method' => $method]);
            if ($ch['status'] === 200) {
                return new JsonResponse($ch['body'], 402);
            }
            if ($ch['status'] === 404) {
                return new JsonResponse($ch['body'], 404);
            }

            return $fail();
        }

        $proof = null;
        $raw = $request->header('X-Payment');
        if ($raw !== null) {
            $decoded = json_decode($raw, true);
            $proof = json_last_error() === JSON_ERROR_NONE ? $decoded : $raw;
        }

        $vreq = ['route' => $route, 'method' => $method, 'nonce' => $proofNonce];
        $payer = $request->header('X-Payment-Payer');
        if ($payer !== null) {
            $vreq['payer'] = $payer;
        }
        if ($proof !== null) {
            $vreq['payment_proof'] = $proof;
        }

        $vr = $client->verify($vreq);
        if ($vr['status'] === 200 && ($vr['body']['allowed'] ?? null) === true) {
            return $next($request);
        }
        if ($vr['status'] === 402) {
            return new JsonResponse($vr['body'], 402);
        }

        return $fail();
    }
}

File: aspnetcore/src/X402Middleware.cs
using System;
using System.Text;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Http;

namespace X402.AspNetCore;

/// <summary>
/// ASP.NET Core middleware that gates a route behind an x402 payment. No
/// business / settlement logic lives here: it relays to the platform's signed
/// challenge/verify and FAILS CLOSED — if the platform is unreachable the
/// route returns HTTP 502 and never serves paid content. The downstream
/// pipeline runs ONLY when the platform explicitly allows the request.
///
/// <para>Wire it with <c>app.UseX402("/premium", "0.10")</c>.</para>
/// </summary>
public sealed class X402Middleware
{
    private readonly RequestDelegate _next;

    /// <summary>The gated route path. Empty ⇒ use the incoming request path.</summary>
    private readonly string _route;

    /// <summary>Telemetry hint only; the authoritative price is the platform route.</summary>
    private readonly string _price;

    private readonly PlatformClient _client;

    /// <summary>Resolves config from the environment and builds a real HTTP client.</summary>
    public X402Middleware(RequestDelegate next, string route, string price)
        : this(next, route, price, new PlatformClient(X402Config.Resolve(null)))
    {
    }

    /// <summary>Injectable client (tests / custom config).</summary>
    public X402Middleware(RequestDelegate next, string route, string price, PlatformClient client)
    {
        _next = next;
        _route = route ?? string.Empty;
        _price = price;
        _client = client;
    }

    public string Price => _price;

    public async Task InvokeAsync(HttpContext ctx)
    {
        var routeKey = _route.Length == 0 ? ctx.Request.Path.ToString() : _route;
        var method = ctx.Request.Method;

        var proofNonce = ctx.Request.Headers["X-Payment-Nonce"].ToString();

        if (string.IsNullOrEmpty(proofNonce))
        {
            var ch = _client.Challenge(routeKey, method);
            switch (ch.Status)
            {
                case 200:
                    await Relay(ctx, 402, ch.Body);
                    return;
                case 404:
                    await Relay(ctx, 404, "{\"error\":\"no_such_route\"}");
                    return;
                default:
                    await FailClosed(ctx);
                    return;
            }
        }

        var proof = ctx.Request.Headers["X-Payment"].ToString();
        var payer = ctx.Request.Headers["X-Payment-Payer"].ToString();

        var vr = _client.Verify(routeKey, method, proofNonce, payer, proof);

        if (vr.Status == 200 && vr.JsonBoolean("allowed"))
        {
            await _next(ctx);
            return;
        }
        if (vr.Status == 402)
        {
            await Relay(ctx, 402, vr.Body);
            return;
        }
        await FailClosed(ctx);
    }

    private static async Task Relay(HttpContext ctx, int status, string body)
    {
        ctx.Response.StatusCode = status;
        ctx.Response.ContentType = "application/json";
        var payload = string.IsNullOrEmpty(body) ? "{}" : body;
        await ctx.Response.Body.WriteAsync(Encoding.UTF8.GetBytes(payload));
    }

    private static Task FailClosed(HttpContext ctx)
        => Relay(ctx, 502, "{\"error\":\"x402_platform_unavailable\"}");
}
